### ISSUE TITLE
Consider adding getifaddrs, getaddrinfo, freeaddrinfo

### ISSUE THREAD

**Issue by psmedley** [bitwiseworks/libcx#37]
The above are useful to have... LGPL implementations are in Samba:
getifaddrs 
https://trac.netlabs.org/samba/browser/trunk/server/lib/replace/getifaddrs.c (HAVE_IFACE_AIX codepath is used for OS/2)
& https://trac.netlabs.org/samba/browser/trunk/server/lib/replace/getaddrinfo.c

**Comment by dmik** [comment]
There is an implementation of getaddrinfo/getnameinfo by KO Myung-Hun, at least as part of VLC (src/os2/getaddrinfo.c). We should borrow it from there.

**Comment by SilvanScherrer** [comment]
while at it please also include https://github.com/bitwiseworks/samba-os2/commit/101a923008735b6f29b3dafb37c2dc2185f052b5 this is also borrowed from Ko Myung-Hun

**Comment by dmik** [comment]
I'm a bit confused. For `getaddrinfo` and friends we end up with at least three implementations (found the third one by accident):

1. From Samba, pointed to by @psmedley.
2. From @komh, 
https://github.com/komh/os2compat/blob/0c148525daac9f296180d83f45386f7c1f2affa7/network/getaddrinfo.c
3. From rsync sources, https://rsync.samba.org/doxygen/head/getaddrinfo_8c-source.html

All of them have minor differences. I'm not going to give it a deep analysis right now. I will pick up the simplest & cleanest implementation (by @komh) and then we will enhance/fix it as needed (perhaps borrowing from other sources).

**Comment by dmik** [comment]
Just for the record: one of the key differences of `gethostinfo` compared to the old-style `gethostbyname` is its reentrancy. Neither of the above implementations guarantee it. Another key feature is transparent IPv6 support but this we lack and will return EAFNOSUPPORT in appropriate cases.

**Comment by dmik** [comment]
Also, I see at least one bug in the Samba's implementation in `getnameinfo`: it returns either the host name or the service name but according to http://man7.org/linux/man-pages/man3/getnameinfo.3.html should return both.

**Comment by dmik** [comment]
I've added `getaddrinfo` and `getnameinfo` (and a test case) to LIBCx. Note that we'll gather all  our network extensions in a single include file, `libcx/net.h`. The `getaddrinfo` and `getnameinfo` originally belong to `netdb.h` but since this file is provided by kLIBC, we can't change it from LIBCx. Once LIBCx is merged into kLIBC, we will move the declarations to the right headers.

Note that a closer look revealed that the KoMH (VLC) code provides a dummy `getnameinfo` implementation that simply converts the address to its string representation (x.x.x.x) and doesn't do an actual host lookup. The implementation provided by samba is much better as it does the actual lookup. I will merge it in.

**Comment by dmik** [comment]
While testing `getnameinfo`, I looked even closer at the Samba and KoMH (VLC) versions and found out that the  Samba version of `getaddrinfo` is actually more sophisticated too. For instance, the VLC version only returns one address for `bitwiseworks.com` while the Samba version returns all four (just like nslookup). However, the VLC version returns two entries per each address (one for SOCK_STREAM/IPPROTO_TCP and one for SOCK_DGRAM/IPPROTO_UDP). And the Samba version returns one address entry where the socket type is either copied from the hints argument or set to SOCK_STREAM if not specified and protocol is always copied from the hints (will be 0 if not specified). OTOH the values returned by `getaddrinfo` are to be passed to the socket should be fine most of the time and passing 0 as protocol is normally fine as each socket type usually implies a single protocol...

This is getting a bit too hairy. Too many things to consider, too many standards to check. Sigh.

I tend to replace the VLC version with the Samba version completely, though.

**Comment by dmik** [comment]
Hmm, the AIX implementation of getifaddrs from Samba returns `Can't assign requested address` (errno 49) here. Trying to figure out why.

**Comment by dmik** [comment]
Okay, I figured out, it was because I took the upstream version w/o OS/2 patches from https://trac.netlabs.org/samba/browser/trunk/server/lib/replace/getifaddrs.c. Applying the patches makes it work I get double entries in the returned list. Need to sort that out.

**Comment by dmik** [comment]
Ok, I found out what's wrong. Actually, all is wrong. The Samba version is wrong and the OS/2 patch is wrong too. The Samba version does not check for `ifr_addr.sin_family` in the returned array of `ifreq` structures in reply to SIOCGIFCONF when performing SIOCGIFADDR, SIOCGIFFLAGS and SIOCGIFNETMASK requests later. However, the SIOCGIFADDR and SIOCGIFNETMASK requests are only valid for AF_INET addresses (see e.g. https://linux.die.net/man/7/netdevice). And on OS/2 the array of `ifreq` contains not only AF_INET entries but also AF_LINK ones (and these are actuallly`sockaddr_dl` structures, not `sockaddr_in`). An AF_LINK entry is returned for each interface existing on the link level (i.e. lan0-lan7 and lo in case of OS/2). An AF_INET entry is only returned for configured interfaces with IP addresses (usually, lan0 and lo). Hence, configured interfaces appear twice in the array (first as AF_LINK and second as AF_INET). And of course performing SIOCGIFADDR on an AF_LINK-only entry returns EADDRNOTAVAIL (49). The above OS/2 patch to the wrong Samba code is therefore quite a hackish workaround.

I will redo it properly (actually almost redone -- needs some polishing). It makes sense to apply it to our Samba port later (or use LIBCx there instead).

**Comment by dmik** [comment]
Committed everything including my fixes. Note that now `getifaddrs` correctly returns AF_LINK entries in addition to AF_INET ones (it would simply drop them in the original Samba version with the original OS/2 patch, making AF_INET ones to double for configured interfaces). These entries are rather useful as they e.g. allow to obtain the MAC address of the lanX interface.

**Comment by dmik** [comment]
I added a nice extension to `getifaddrs` so that it also returns broadcast/P2P addresses for the interface (according to IFF_BROADCAST and IFF_POINTOPOINT flags).

And the last bit, `if_nameindex` and `if_nametoindex` family APIs, are also there. Now we are packed and set. The only missing thing is actual reentrancy of the calls. If I knew that `gethostbyname` family APIs in the OS/2 TCP/IP 4.21 DLL are reentrant, we would not need that. But unfortunately I don't. If anyone has such a knowledge, please share. If not, I will use a fast and lightweight `_fmutex` to guard `get*by*` calls.

**Comment by dmik** [comment]
Well, on the second thought, reentrancy is actually not an easy task. Imagine if two threads request some name resolution in parallel and and the first one gets delayed due to a slow link to the underlying name server: this will block the second request completely, even if it's fast. I guess that the only way to properly solve it is to implement name resolving from scratch as even having a pool of threads won't help if `gethostbyname` is not thread safe or reentrant inside. We should check if we could use the `bind` library from *nix for this but it's clearly besides the scope of this task. I will leave it as is for now.

**Comment by dmik** [comment]
On the third thought, I still decided to use a mutex. The reason is that if `gethostbyname` and friends are in fact non-reentrant in the OS/2 TCP/IP DLL, then calling them from several threads will result in unexpected behavior or crash anyway. WIth the mutex, the second thread will just have to wait. It's not true reentrancy of course, but at least it will let to avoid unexpected behavior. This will be documented in README.OS2 of course.

I think it's enough for this ticket. I will give it a bit more testing against Qt (within https://github.com/bitwiseworks/qtbase-os2/issues/35) and then release a new LIBCx RPM.

**Comment by dmik** [comment]
LIBCx 0.6.3 with all the above stuff is just released as RPM to `exp`. Feel free to test.
